### PR TITLE
chore(website): [playground] correct issues with libs

### LINKED ISSUE #5142: Docs: Playground not emitting some type-aware lint errors
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

It seems some type-aware rules are not emitting any errors in the playground.

Setting `"@typescript-eslint/no-base-to-string": "error"`, and writing `"" + {}` doesn't trigger errors.

[Playground](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=ETAEGpQbwXwbgFBA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oCMBDZRLXxdk+KkwDm6KImjQO0SODABfECqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

Floating promises don't trigger errors.

[Playground](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=IYZwngdgxgBAZgV2gFwJYHsL3egFAShgG8YAnAU2QVKwEYBuGAXwChEUMsAjYUg4ljGx589FqyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

Type-unaware rules do work.

Maybe I'm doing something wrong? It's at least confusing that configurations don't immediately trigger errors.

### Affected URL(s)

https://typescript-eslint.io/play

### PR DESCRIPTION
**Pull request by armano2**
## PR Checklist

-   [x] Addresses an existing open issue: fixes #5142
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Correct issue with loading libs between editor and code, "lib" resolution imitates to behavior 

### Changes

- editor and linting is using now exact copy of typescript definitions.
  there is no additional calls to cdn to retrieve type definitions, unless sandbox is not exposing types (old ts versions) 
- `lib`, options can be provided in tsconfig and its going to be properly used
- parsing and validation of tsconfig is done by typescript utils
- ctrl+click on types defined in libs is now working correctly

  lib: es6
  image
  lib: esnext
  image

### Test cases
- [Playground main](https://typescript-eslint.io/play/#ts=4.7.1-rc&sourceType=module&code=PTAEFpK6ZhYAUCUBnALgJwJYGM3gCMB7IgGwFMBDAO3HIA8AHDclFLI6lRZWPvxIiwAzUAApMAV3KgAZLNABGUAGolASlABvUAF9EAdwAWWCuOGVSKGQB8boAESUHqxwQeat+hGPoAuUCICACtyPE0AXgA+UClbewsreNB6UAB+WIxpUADE6wBuQSQwflKoHjBqIkJKa3A0avRsagBzCogy0qKHFzUvfNBQZEAZclAqmrqG8CasVqLeTv528eFSIko0WZbwZiIAWyxrbmKOxdgi2oBPahxQYUkbzc47kjFPUBY0SQxqJQHvBiMIgYNB3B54Di-AiUDBvbSIQbCV7qAYjMbVVbrTatHYYfaHVigRDeRAANyIWAAJmNyAZQAAFPEHaxiOHRbR6FFDMCjFZrDZbXH4o5EhCIQHA0H3R6Q0A4FgbciM4XkOFaBEfchfH40unK5mqtkxHS6LnIQCg5JrtdRyNTiUV5VQ0EqmQS3qieejwJiBTjdgbjuKmJKwTLnuhJMJhIo1RqcJx0KBGKAIrqGa6WUaOaaPaBedUHpJrJTwKSYccSQgJSDQFcbqGIeGvlGAEyxhCDSgGShYUE2vUZw2RY2c3OWrs952U+1iqvBmvSxu-IsulViRgBfUEgA8DwA1lUDNQop5dNy817C8XS+Wiiutyz++m11mTeozWBLYxasWikGgTWda3IuTy-JYpDtoMgzxlwoIACSUCmcoKs6D6qh+oDjmwWAtDa04dlBMGJnBBBIY6ipoe655YewuG2kUUFygmoIANqUAANKABAALpIROvbPgaAB04FiCxCGcSR3EYV+P62rERCCQSImkBB6gaoMaLjFeto3hgFZFCI4jkahg5vKe575uAzIriW-oEscximDIYgmauBrmdoZ5adUNnXvZRyIK5KHuW6mgZHEOR3JY1gYVZfm6QFrBBW5lGaHYmTSHFXoJXZg7HGIACEqVme+ubxYctlCgGRSUmEpAwjIIGypQOA4OQjBoCgABqFKUgAwgAQq5BABFm5JUiiiCte1nU9X1Q3BU6oWxeVOWVf5+WIEAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs+SAIZN6AD2LQUySnyaoMAbXDYciaND7RIAGhWqsq7JCHx4fAO4BlTswDm6MADNTyRHqPHT5iwDlYALY86o4u8G4enlDelv5mQjxIAPI8AFaI3KGuiPrYAL65ALqROAQk5FQ09MiMLKxMfLRO5kL49rSSfAGUbgpQ6praJVBlpMgU1HQMzGwNtN3IsG4AJh2aCyiOkANaurl4RGMTVdN1c7BMS4irAG5C0H3bGrvDB+XjlVM1M-WNPEJuWj4RocKhMBwYJ6DSAqAp5IA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)
- [Playground this branch](https://deploy-preview-5162--typescript-eslint.netlify.app/play/#ts=4.7.1-rc&sourceType=module&code=PTAEFpK6ZhYAUCUBnALgJwJYGM3gCMB7IgGwFMBDAO3HIA8AHDclFLI6lRZWPvxIiwAzUAApMAV3KgAZLNABGUAGolASlABvUAF9EAdwAWWCuOGVSKGQB8boAESUHqxwQeat+hGPoAuUCICACtyPE0AXgA+UClbewsreNB6UAB+WIxpUADE6wBuQSQwflKoHjBqIkJKa3A0avRsagBzCogy0qKHFzUvfNBQZEAZclAqmrqG8CasVqLeTv528eFSIko0WZbwZiIAWyxrbmKOxdgi2oBPahxQYUkbzc47kjFPUBY0SQxqJQHvBiMIgYNB3B54Di-AiUDBvbSIQbCV7qAYjMbVVbrTatHYYfaHVigRDeRAANyIWAAJmNyAZQAAFPEHaxiOHRbR6FFDMCjFZrDZbXH4o5EhCIQHA0H3R6Q0A4FgbciM4XkOFaBEfchfH40unK5mqtkxHS6LnIQCg5JrtdRyNTiUV5VQ0EqmQS3qieejwJiBTjdgbjuKmJKwTLnuhJMJhIo1RqcJx0KBGKAIrqGa6WUaOaaPaBedUHpJrJTwKSYccSQgJSDQFcbqGIeGvlGAEyxhCDSgGShYUE2vUZw2RY2c3OWrs952U+1iqvBmvSxu-IsulViRgBfUEgA8DwA1lUDNQop5dNy817C8XS+Wiiutyz++m11mTeozWBLYxasWikGgTWda3IuTy-JYpDtoMgzxlwoIACSUCmcoKs6D6qh+oDjmwWAtDa04dlBMGJnBBBIY6ipoe655YewuG2kUUFygmoIANqUAANKABAALpIROvbPgaAB04FiCxCGcSR3EYV+P62rERCCQSImkBB6gaoMaLjFeto3hgFbzCUZznCc4zMiuJb+gSxwLMZ0BFCI4jkahg5vKe575uA5nXlZRyGCYZhiM5q4Gm52hnlp1TebpvmsIgQUoSFbqaBkcQ5HcljWBhnnRZZg7HAlTpJSymh2Jk0jZV6uVCgG8UAITBZR765jlhwWTV1lFJSYSkDCMggbKlA4Dg5CMGgKAAGoUpSADCABCQUEAEWbklSKKIENI1jZN03zYVFGDpVZltT5+WIEAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs+SAIZN6AD2LQUySnyaoMAbXDYciaND7RIAGhWqsq7JCHx4fAO4BlTswDm6MADNTyRHqPHT5iwDlYALY86o4u8G4enlDelv5mQjxIAPI8AFaI3KGuiPrYAL65ALqROAQk5FQ09MiMLKxMfLRO5kL49rSSfAGUbgpQ6praJVBlpMgU1HQMzGwNtN3IsG4AJh2aCyiOkANaurl4RGMTVdN1c7BMS4irAG5C0H3bGrvDB+XjlVM1M-WNPEJuWj4RocKhMBwYJ6DSAqAp5IA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

### CODE REVIEW

File: packages/website/src/components/editor/config.ts [hunk 1]
```diff
@@ -2,17 +2,30 @@ import type Monaco from 'monaco-editor';
...
+  const libs = Array.isArray(tsConfig.lib) ? tsConfig.lib : ['es6', 'dom'];
+
   return {
-    noResolve: true,
     // ts and monaco has different type as monaco types are not changing base on ts version
     target: window.ts.ScriptTarget.ESNext as number,
     module: window.ts.ModuleKind.ESNext as number,
     ...tsConfig,
     jsx: jsx ? window.ts.JsxEmit.Preserve : window.ts.JsxEmit.None,
+    lib: libs.map(item => window.ts.libMap.get(String(item)) ?? String(item)),
```
Comment (armano2) [inline packages/website/src/components/editor/config.ts:28]: sandbox is not correctly resolving lib names

File: packages/website/src/components/editor/useSandboxServices.ts [hunk 2]
```diff
@@ -77,19 +78,31 @@ export const useSandboxServices = (
...
-            lib: libs.filter(item => !item.includes('.')),
-          },
-          props.ts,
-          true,
-          window.ts,
-        );
-        const system = sandboxInstance.tsvfs.createSystem(libMap);
+        let libEntries: Map<string, string> | undefined;
+        const worker = await sandboxInstance.getWorkerProcess();
+        if (worker.getLibFiles) {
```
Comment (armano2) [inline packages/website/src/components/editor/useSandboxServices.ts:83]: not present for sandbox with ts < 3.7

File: packages/website/src/components/editor/useSandboxServices.ts [hunk 3]
```diff
@@ -77,19 +78,31 @@ export const useSandboxServices = (
...
+          libEntries = new Map(
+            Object.entries((await worker.getLibFiles?.()) ?? {}),
+          );
+        } else {
+          // for some older version of playground we do not have definitions available
+          libEntries = await sandboxInstance.tsvfs.createDefaultMapFromCDN(
```
Comment (armano2) [inline packages/website/src/components/editor/useSandboxServices.ts:89]: fallback for older versions of sandbox

File: packages/website/src/components/editor/config.ts [hunk 4]
```diff
@@ -2,18 +2,38 @@ import type Monaco from 'monaco-editor';
...
+type StringMap = Map<string, string>;
+declare module 'typescript' {
+  /**
+   * Map of available libraries
+   *
+   * The key is the key used in compilerOptions.lib
+   * The value is the file name
+   */
+  const libMap: StringMap;
+}
```
Comment (bradzacher) [inline packages/website/src/components/editor/config.ts:14]: can we please put this in `typings/typescript.d.ts`?

It's nice to make it clear what "private" surface we expose.

**Comment by armano2** [comment]
@bradzacher suggested change is applied

File: packages/website/src/components/editor/config.ts [hunk 5]
```diff
@@ -6,14 +6,21 @@ export function createCompilerOptions(
...
-    noResolve: true,
+  const config: Monaco.languages.typescript.CompilerOptions = {
     // ts and monaco has different type as monaco types are not changing base on ts version
     target: window.ts.ScriptTarget.ESNext as number,
     module: window.ts.ModuleKind.ESNext as number,
     ...tsConfig,
     jsx: jsx ? window.ts.JsxEmit.Preserve : window.ts.JsxEmit.None,
   };
+
+  const libs = Array.isArray(tsConfig.lib) ? tsConfig.lib : ['es6', 'dom'];
```
Comment (armano2) [inline packages/website/src/components/editor/config.ts:17]: this is not fully correct, we should deduce libs if not provided from target and module :(

**Review by bradzacher — APPROVED** [review]
looks-good-to-me

**Pull request merged**